src/networking.c

```diff
@@ -3905,20 +3887,13 @@ void unblockPostponedClients() {
  * failover procedure implemented by CLUSTER FAILOVER.
  *
  * The function always succeed, even if there is already a pause in progress.
- * In such a case, the duration is set to the maximum and new end time and the
- * type is set to the more restrictive type of pause. */
-void pauseClients(pause_purpose purpose, mstime_t end, pause_type type) {
+ * The new values of a given 'purpose' will override the old ones */
+void pauseServices(pause_purpose purpose, mstime_t end, uint32_t bitmask) {
     /* Manage pause type and end time per pause purpose. */
-    if (server.client_pause_per_purpose[purpose] == NULL) {
-        server.client_pause_per_purpose[purpose] = zmalloc(sizeof(pause_event));
-        server.client_pause_per_purpose[purpose]->type = type;
-        server.client_pause_per_purpose[purpose]->end = end;
-    } else {
-        pause_event *p = server.client_pause_per_purpose[purpose];
-        p->type = max(p->type, type);
-        p->end = max(p->end, end);
-    }
```

Comment:
i think there was an intention that if a new pause was started for a purpose that is already active, we take the max of the time (and effects), rather than override it with a shorter / weaker one.

Reply:
let's look at it from the user's perspective.
till now, issuing `CLIENT PAUSE 10` and then `CLIENT PAUSE 2` would have kept the 10 (it was like that even in redis 3.0), so changing it could be a breaking change.

Reply:
>Suppose the a client entity has new better shorter-evaluation to the duration , then it is stuck with old value, unless it is unpause+pause again.

I think this gets into why is a timer being used here instead of some cron that eventually unpauses clients. Almost all timeouts are user supplied or artificial limits that aren't going to change. Lowering timeouts doesn't really make sense to me, which is why I think all the pauses should follow the current logic.

Reply:
I know that when I saw this code for the first time it looked little confusing, but the context was clear, which is, pause-clients. Now that the infrastructure become more generic (i.e. pause-actions) it makes less sense to keep such additional logic in the core, but let the entity to decide from outside.

Reply:
> Now that the infrastructure become more generic (i.e. pause-actions) it makes less sense to keep such additional logic in the core, but let the entity to decide from outside.

It's not more generic though, it's *more* tightly coupled than before. Before we had a single generic mechanism that set a timeout with a behavior (never allow timeouts to shorten). Now we have a different behavior (always set the timeout to X), and custom code that handles one specific case within the core. If the point was to be generic, we should have two separate flags we can pass into pauseActions(), one that only extends the timeout and one that sets it.